mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp

```diff
@@ -22,13 +22,32 @@
 #include "llvm/ADT/SmallBitVector.h"
 #include "llvm/ADT/SmallVector.h"
 #include <algorithm>
+#include <numeric>
 
 using namespace mlir;
 using namespace mlir::linalg;
 
 /// Include the definitions of the copy operation interface.
 #include "mlir/Dialect/Linalg/IR/LinalgInterfaces.cpp.inc"
 
+namespace {
+/// Check linalg generic with single input output has
+/// body that is just a yield op yielding input value.
+static bool bodyIsJustYieldOp(GenericOp genericOp) {
```

Comment:
Could you perhaps use this mechanism instead?
https://github.com/plaidml/tpp-mlir/blob/main/include/TPP/IR/StructuredOpMatcher.h

We can create a PR to upstream this part, so that we can all use it.

@ftynse any objections?

Reply:
> Could you perhaps use this mechanism instead? https://github.com/plaidml/tpp-mlir/blob/main/include/TPP/IR/StructuredOpMatcher.h
> 
> We can create a PR to upstream this part, so that we can all use it.
@rengolin please could you elaborate what exactly you mean ? TPP i am not so familiar with and its outside mlir-core.

Reply:
Ping!

Reply:
My suggestion has nothing to do with our project. The code in the file is pure LLVM.

I see that this file has a lot of helper functions `isThis` or `isThat` and this is the state we had in our project 2 years ago. Early last year we moved on to _structured matchers_ (as per the file), and it makes your life much easier because you can describe the pattern you want like a builder design pattern and not like a bunch of if/else/error snippets and helper functions that do pretty much the same thing over and over.

We can just clean up and try to submit a patch to add this and replace the existing mechanism, then you rebase your patch. Or you can _steal_ our code and add to this PR and make your life easier. Either way works for me.

Reply:
"We can just clean up and try to submit a patch to add this " -- No. I will do it.
But you have to explain a bit more in detail, for instance is there any existing code in MLIR core that uses this approach. The functions in https://github.com/plaidml/tpp-mlir/blob/main/include/TPP/IR/StructuredOpMatcher.h seem duplicating what (e.g. NumLoops) are already straight-away in MLIR-linalg. 
Also, what going to happen to ContractionOps and existing generic-to-named OpInterfaces already written out. Port all of them? 
I want to make the generic-to-named more complete (e.g. convolution, pooling ... is missing). I dont see how another approach will make it faster. Maybe i am missing something but based on what i see i think current approach is not more verbose

Reply:
> Also, what is going to happen to ContractionOps and existing generic-to-named OpInterfaces which are already written out. Port all of them?

Eventually yes. But I wouldn't try to push that as a blocker.

> I want to make the generic-to-named more complete (e.g. convolution, pooling ... is missing). I dont see how another approach will make it faster.

It's not about faster, it's about avoiding the bloat that this conversion is already creating. We had the exact same problem that we had dozens and dozens of helper functions. I think there's already prior art in MLIR (where we took it from), so if you want to use that, probably better.

> The main challenge in generic-to-named is distilling and abstracting semantics

Yup, and that is currently spread across functions that are basically identical to each other except one thing.

If you can at least parametrize / templatize the existing calls, that would already help a lot. `isCopy`, `isFill`, `isBroadcast` all call the same checks, over and over.